Update Docs for 1.0 Release

micahlee opened the pull request:
Connected to #5

izgeri commented on docs-content/index.html.md.erb:
```diff
@@ -69,12 +69,131 @@ The following table provides version and version-support information about Cyber
+  * Conjur Open Source (v5.x)
+  * Hosted Conjur. You can quickly get a hosted Conjur instance running for evaluation purposes. Visit the [hosted Conjur page](https://www.conjur.org/get-started/try-conjur.html) to sign up.  
+  
+  ???are there any limitation to the version numbers?  v4.x is very broad.... is it v4.8 or .9?? 
```
We've been testing with 4.9.x, but I'm not sure how many clients are on an older version. Can you check with Dan?

perrygold replied:
Fixed per Dan's reply: (4.9.12.0 and later)
4.9.12.0 — the latest version of Conjur 4. This integration is targeted existing CyberArk customers that are new to Conjur and have or are about to purchase Conjur Enterprise

izgeri commented on docs-content/index.html.md.erb:
```diff
@@ -69,12 +69,131 @@ The following table provides version and version-support information about Cyber
+
+  * Conjur Enterprise (v4.x)
+  * Conjur Open Source (v5.x)
+  * Hosted Conjur. You can quickly get a hosted Conjur instance running for evaluation purposes. Visit the [hosted Conjur page](https://www.conjur.org/get-started/try-conjur.html) to sign up.  
```
Hosted Conjur should be a sub-bullet of Conjur Open Source - it's a hosted version of the OSS product.

perrygold replied:
fixed

izgeri commented on docs-content/index.html.md.erb:
```diff
@@ -69,12 +69,131 @@ The following table provides version and version-support information about Cyber
+
+     
++ To configure the PCF tile, you must know the following information about your Conjur appliance:    
+  * Conjur Account ID, which is specified when the Conjur appliance is created and configured. For hosted Conjur, the account is usually an email address that you specify on the web form.
```
I would just say Conjur Account here and drop the ID - at first I thought you were talking about a username bc of it

perrygold replied:
fixed

izgeri commented on docs-content/index.html.md.erb:
```diff
@@ -69,12 +69,131 @@ The following table provides version and version-support information about Cyber
+  * URL to the Conjur appliance.  
+  * Valid Conjur user name and associated API Key for a user that has the ability to update and create PCF-related policy. The Service Broker uses these credentials to add and remove host identities as you deploy and remove PCF applications. See [Recommendations](#recommendations).  
+  *  Conjur policy namespace for PCF, if available. For production Conjur appliances, CyberArk strongly recommends  a dedicated PCF namespace;  otherwise, the Conjur root policy is the default.  See [Recommendations](#recommendations) for more information. 
+  *  Conjur version. Either v4 for Enterprise Conjur or v5 for Open Source or hosted Conjur. 
```
The default is 5

perrygold replied:
removed the "v's" .   want to mention  the default in the install section, not here.

izgeri commented on docs-content/index.html.md.erb:
```diff
@@ -69,12 +69,131 @@ The following table provides version and version-support information about Cyber
+  * Conjur Account ID, which is specified when the Conjur appliance is created and configured. For hosted Conjur, the account is usually an email address that you specify on the web form.
+  * URL to the Conjur appliance.  
+  * Valid Conjur user name and associated API Key for a user that has the ability to update and create PCF-related policy. The Service Broker uses these credentials to add and remove host identities as you deploy and remove PCF applications. See [Recommendations](#recommendations).  
+  *  Conjur policy namespace for PCF, if available. For production Conjur appliances, CyberArk strongly recommends  a dedicated PCF namespace;  otherwise, the Conjur root policy is the default.  See [Recommendations](#recommendations) for more information. 
```
" (optional)" instead of ", if available", please

perrygold replied:
ok. i was worried Pivotal would ask us to remove it out of the Required section! but let's try it.

izgeri commented on docs-content/index.html.md.erb:
```diff
@@ -69,12 +69,131 @@ The following table provides version and version-support information about Cyber
+
+  *  The tile installation requires a copy and paste of the x509 certificate that was created when Conjur was initiated. The certificate is stored in a `.pem` file that was created by the `conjur init` command. The default path is `/conjur_install_directory/conjur-account_id.pem`. 
+  
+  *  The Service Broker needs a Conjur host-factory to use when adding host identities. See [Recommendations](#recommendations) for examples.
```
"host-factory" => "Host Factory"

perrygold replied:
fixed

izgeri commented on docs-content/index.html.md.erb:
```diff
@@ -69,12 +69,131 @@ The following table provides version and version-support information about Cyber
+  * The secrets, host identities, and  access control policies for your PCF-deployed applications are isolated from other resources stored in Conjur.
+  *  Conjur access privileges required by the Service Broker,  DevOps teams, and your applications can be limited in scope to the PCF namespace.     
+    * The Service Broker adds and removes host identities to Conjur as you deploy and remove applications.  If you provide a PCF namespace in the tile configuration, those hosts are added into that namespace. To comply with least privilege goals, you can create an  administrative user name for the Service Broker within the dedicated PCF namespace. Otherwise, the Service Broker needs update privil ...
+   * PCF DevOps personnel need access to Conjur to declare application secrets and grant privileges to the applications to access those secrets. Recommended best practice is to create a group of users within the dedicated PCF namespace for this purpose. 
```
I'm not sure what this bullet really adds to this list

izgeri replied:
If you rewrite the bullet above as I've suggested, you might want to keep this one - but I think we need to be clearer about what point we're trying to get across.

Are we trying to say that we can limit the Conjur users who have access to update policies for applications deployed to PCF? Because that's valid and useful information. But that presumably will be accomplished by defining even finer-grained permissions than the `pcf-admin-group` who owns the PCF policy namespace.

Really, this namespace facilitates giving ownership access to manage Conjur policies for PCF applications to a limited set of Conjur users, with even finer-grained permissions possible within the policy namespace itself.

perrygold replied:
I do want to specifically state that the SB and PCF devs need privileges to access Conjur, and why.  We dance around these concepts and they should be understood up front. I shortened up the description considerably and moved it; it was overkill. See what you think.

izgeri commented on docs-content/index.html.md.erb:
```diff
@@ -69,12 +69,131 @@ The following table provides version and version-support information about Cyber
+    
+
+  ??? 
+  where are privileges assigned to role; feels like something is missing?    
```
where did `pcf-admin-broker` come from? this is not needed

where are privileges assigned to which role?

perrygold replied:
there are two reasons where access is needed into the space. I was just trying to make that clearer by using two different users rather than one. (and illustrating the group concept.) that is really least privilege, right?  But ok,    I can make it just one user and simply state that finer control is possible. 
that is really least privilege, right?

perrygold replied:
My role question is this: where is the statement that assigns read-update privileges to the role we are using?  it is implied but I think we should show it. Unless there is something about roles that I don't understand.

izgeri commented on docs-content/index.html.md.erb:
````diff
@@ -69,12 +69,131 @@ The following table provides version and version-support information about Cyber
+3. Load the policy into Conjur under the root policy, as follows:
+   
+   ```
+   $ conjur policy load root your-pcf-policy-file.yml
````
this is not the syntax for v4

probably would be something like
`conjur policy load --as-group security_admin your-pcf-policy-file.yml`

perrygold replied:
do you mean, is the syntax different for v4 and v5? I wrote it that way, two diff syntaxes.

izgeri commented on docs-content/index.html.md.erb:
```diff
@@ -69,12 +69,131 @@ The following table provides version and version-support information about Cyber
+   
+3. In the CyberArk Conjur Service Broker for PCF tile configuration: 
+  *  For **Conjur Policy**, use the policy id (`pcf` in the above example)   
+  *  For **Conjur Login**, use one of the user ids declared in the pcf namespace. 
```
use a member of the group that owns the Conjur Policy namespace

izgeri replied:
For Conjur API Key, use that user's API key - which can be retrieved by the admin who created that role using:

`conjur user rotate_api_key -u pcf-admin`

perrygold replied:
done

izgeri commented on docs-content/installing.html.md.erb:
"have the ability to update Conjur poliy" => "have `create` and `update` privileges on the Conjur policy"
"must have privileges on the Conjur root policy" => "must have `create` and `update` privileges on the Conjur root policy"

perrygold replied:
done

izgeri commented on docs-content/installing.html.md.erb:
Select 5 for Open Source Conjur (including eval hosted Conjur)

perrygold replied:
done

izgeri commented on docs-content/using.html.md.erb:
```diff
@@ -26,48 +30,80 @@ You can push the same application to multiple spaces. After preparing the applic
   1. Create a file named `secrets.yml`.
  
-  2. Add entries to the file to define the secrets that the application will fetch from the Conjur appliance.
+  2. Add entries to the file to define the secrets that the application will fetch. Secrets to fetch from Conjur are specified with a !var tag and a pathname indicating the Conjur policy path.  
```
not all secrets use a `!var` tag (for example, `!file`). we write the general label as `!<tag>` in the summon docs

perrygold replied:
? are the secrets that Summon will fetch from Conjur !var tags? that's what I'm saying here. Can Summon fetch a secret from a !file tag?

perrygold replied:
you answered that below, saying yes it can.  there is still something wrong with this sentence, I think (in the case of !file, is the pathname a conjur pathname? prob. not)

izgeri commented on docs-content/using.html.md.erb:
```diff
@@ -26,48 +30,80 @@ You can push the same application to multiple spaces. After preparing the applic
  
   2. Store the `secrets.yml` file in the root folder of your application. 
+  
+  4. Create Conjur policy that declares the secrets as Conjur variables and grants access to them for users.   [A later step](#update-policy) grants access to the PCF application.  
```
"grants access to them for users (if desired)" <-- they don't *have* to grant access to users

izgeri commented on docs-content/using.html.md.erb:
````diff
@@ -26,48 +30,80 @@ You can push the same application to multiple spaces. After preparing the applic
+  5. Load the policy into Conjur under the PCF namespace. For example, the following command loads a policy file named `my-app.yml` into a Conjur namespace named `pcf`. 
+  
+  ```
+  $ conjur policy load pcf my-app.yml
````
this is v5 syntax. in v4 you would add an `!include my-app.yml` to the `pcf` policy declaration and reload the pcf policy with `conjur policy load --as-group security_admin pcf.yml`

we'll probably want to review the overall flow of this in the document to make sure it's consistent

perrygold replied:
hm. Definitely agree calling out different v4 and v5 syntax.  However,  the include and loading for v4 seems like a detail that experienced users would figure out on their own and we are complicating the example unnecessarily here.  Reconsider if we couldn't just say  (for v4)  conjur policy load --as-group security_admin my-app.yml
(but I did it as you state above for now)

izgeri commented on docs-content/using.html.md.erb:
````diff
@@ -26,48 +30,80 @@ You can push the same application to multiple spaces. After preparing the applic
 ```
 
-Lines 1 and 2 specify variables with pathnames containing the secret.
+Lines 1 and 2 specify variables with pathnames containing the secret. In this case, the pathnames are Conjur policy paths, using the PCF namespace as the first component, followed by other ids in the policy hierarchy. 
````
this is assuming the app policy `my-app` has been loaded into the `pcf` policy namespace, and it's worth calling that out

perrygold replied:
fixed

izgeri commented on docs-content/using.html.md.erb:
```diff
@@ -26,48 +30,80 @@ You can push the same application to multiple spaces. After preparing the applic
 Line 4 specifies a variable that is a file containing the secret.
 
-In the context of a PCF environment, the above example could produce the following results: 
+In the context of a PCF environment, the above example could produce the following results, with the first two values retrieved from Conjur: 
```
the contents of the file would also be retrieved from Conjur.

perrygold replied:
ok, answers question above. fixed

izgeri commented on docs-content/using.html.md.erb:
```diff
@@ -82,55 +118,62 @@ Add the instance-name from the previous step to the list of services in the appl
  1. Log into the Conjur appliance.   
  
- 2. Update Conjur policy to add the new host ID with read and execute privilege on a set of secrets. 
+ 2. Update Conjur policy to add the new host ID with read and execute privilege on a set of secrets, and load the update. 
```
how about just "Update Conjur policy to grant read and execute privileges to the new host ID on a set of secrets"

perrygold replied:
done

jtuttle commented on docs-content/using.html.md.erb:
```diff
@@ -82,55 +118,62 @@ Add the instance-name from the previous step to the list of services in the appl
 The binding process generates a binding ID that becomes the Conjur host ID, used as a unique application identity for this application running in this space.
 
-The ID is stored in the `authn_login` string found under the VCAP_SERVICES environment variable. The `authn_login` string has the format `host/<id>`. For example:   
+The ID is stored in the `authn_login` string found under the VCAP_SERVICES environment variable. The `authn_login` string has the format `host/id`. For example:   
```
Maybe write "The `authn_login` string has the format `host/binding_id`", just to be a bit more explicit?

perrygold replied:
done

izgeri commented on docs-content/index.html.md.erb:
```diff
@@ -69,12 +69,131 @@ The following table provides version and version-support information about Cyber
+ 
+  * The secrets, host identities, and  access control policies for your PCF-deployed applications are isolated from other resources stored in Conjur.
+  *  Conjur access privileges required by the Service Broker,  DevOps teams, and your applications can be limited in scope to the PCF namespace.     
+    * The Service Broker adds and removes host identities to Conjur as you deploy and remove applications.  If you provide a PCF namespace in the tile configuration, those hosts are added into that namespace. To comply with least privilege goals, you can create an  administrative user name for the Service Broker within the dedicated PCF namespace. Otherwise, the Service Broker needs update privil ...
```
principles?

jtuttle replied:
Ya know, I almost googled that to see if I had screwed it up but I must have gotten distracted. Thanks. :D

izgeri replied:
Also, it should be "`create` and `update` privileges on the root policy"

This has a very specific meaning in Conjur, and we should be consistent about referring to these _two_ requirements when we talk about the permissions that the SB Conjur user needs.

perrygold replied:
all done . 
the exact location of the least privilege phrase has changed because of previous comments; but I used the phrase instead of "security best practices" ; liked it better.

izgeri commented on docs-content/index.html.md.erb:
```diff
@@ -69,12 +69,131 @@ The following table provides version and version-support information about Cyber
+A dedicated namespace supports security best practices in the following ways: 
+ 
+  * The secrets, host identities, and  access control policies for your PCF-deployed applications are isolated from other resources stored in Conjur.
+  *  Conjur access privileges required by the Service Broker,  DevOps teams, and your applications can be limited in scope to the PCF namespace.     
```
How about just "Access privileges required by the Service Broker will be limited in scope to the PCF namespace."

izgeri commented on docs-content/index.html.md.erb:
```diff
@@ -69,12 +69,131 @@ The following table provides version and version-support information about Cyber
+                                       /* It does not have to be pcf
+      owner: !group pcf-admin-group
+      body:                            /* The body and following layer are 
+       - !layer pcf-apps               /* required for Conjur Enterprise v4; 
```
Also, it looks like there are tabs in this policy - I am not sure this will work, and our demos elsewhere are indented by 2 spaces in each place. Should be the same here for consistency.

perrygold replied:
done. 
added info about naming the layer and h-f.
fixed spacing ; didn't find any tabs so not sure about that.

perrygold commented on docs-content/index.html.md.erb:
```diff
@@ -69,12 +69,131 @@ The following table provides version and version-support information about Cyber
+    * The Service Broker adds and removes host identities to Conjur as you deploy and remove applications.  If you provide a PCF namespace in the tile configuration, those hosts are added into that namespace. To comply with least privilege goals, you can create an  administrative user name for the Service Broker within the dedicated PCF namespace. Otherwise, the Service Broker needs update privil ...
+   * PCF DevOps personnel need access to Conjur to declare application secrets and grant privileges to the applications to access those secrets. Recommended best practice is to create a group of users within the dedicated PCF namespace for this purpose. 
+   
+   *  Conjur Enterprise (v4) requires a host-factory in an application layer. You can define that layer in  the PCF policy, rather than in the root policy.  
```
done

perrygold commented on docs-content/index.html.md.erb:
```diff
@@ -69,12 +69,131 @@ The following table provides version and version-support information about Cyber
+
+* The PCF DevOps teams will be altering policy files in the root space. 
+
+* For integration with Conjur v4, a host-factory named  `apps` is required.  Add the following layer declaration to the root policy:  
```
done

izgeri commented on docs-content/index.html.md.erb:
```diff
@@ -69,11 +69,147 @@ The following table provides version and version-support information about Cyber
+
+  *  The tile installation requires a copy and paste of the x509 certificate that was created when Conjur was initiated. The certificate is stored in a `.pem` file that was created by the `conjur init` command. The default path is `/conjur_install_directory/conjur-account.pem`.
+
+  *  The Service Broker needs a Conjur Host Factory to use when adding host identities. See [Recommendations](#recommendations) for examples.
```
Can we say requires instead of needs? The tile installation will fail if they don't have a host factory, so if this could be clearer in this line too (bc it sounds sort of more like a suggestion, especially when we direct them to the Recommendations section), that would be good

perrygold replied:
changed wording a bit ..check it again

izgeri replied:
ok

izgeri commented on docs-content/index.html.md.erb:
```diff
@@ -69,11 +69,147 @@ The following table provides version and version-support information about Cyber
+
+When creating the Conjur policy that declares users in the PCF space, consider the following:
+
+ * The Service Broker needs `create` and `update` access to the namespace to add and remove host identities.
```
"access to the namespace" => "privileges on the policy namespace"

perrygold replied:
done

izgeri commented on docs-content/index.html.md.erb:
````diff
@@ -69,11 +69,147 @@ The following table provides version and version-support information about Cyber
+ Example Policy Declaring a Dedicated PCF Space:
+
+   ```
+  - !user
````
this user can be deleted from the policy

perrygold replied:
done

izgeri commented on docs-content/index.html.md.erb:
````diff
@@ -69,11 +69,147 @@ The following table provides version and version-support information about Cyber
+        - !host-factory
+            id: pcf-apps
+            layers: [ !layer pcf-apps ]
+   ```
````
I would prefer that the policy be updated so that it is indented consistently with 2 spaces. I can do that if it's easier.

perrygold replied:
done, i think. If still not right, go ahead and correct.

izgeri commented on docs-content/index.html.md.erb:
```diff
@@ -69,11 +69,147 @@ The following table provides version and version-support information about Cyber
+
+3. In the CyberArk Conjur Service Broker for PCF tile configuration:
+  *  For **PCF Conjur Policy Namespace**, use the policy id (`pcf` in the above example)
+  *  For **Conjur Login**, use a member of the group that owns the PCF policy namespace.
```
"use a Conjur user who is a member"
to be clear we are not talking about a host or a group, etc

perrygold replied:
done

izgeri commented on docs-content/installing.html.md.erb:
```diff
@@ -42,18 +42,29 @@ This tab configures communication between the Service Broker and a Conjur applia
 ![Conjur Service Broker Configuration tab](service_broker_config_screen.png)
```
you will need a new screenshot, since we reordered things and changed the labels :)

perrygold replied:
done

izgeri commented on docs-content/installing.html.md.erb:
"in the dedicated PCF policy space" => "on the dedicated PCF policy ID"

perrygold replied:
done

izgeri commented on docs-content/installing.html.md.erb:
I'm not sure if it's worth editing this to allow for Enterprise v5 users - that's coming before too long.

perrygold replied:
it's true for now, so let's leave it. We will have other edits to make when the time comes.

perrygold replied:
Also,  after checking with Ty, I am changing all of the Conjur Enterprise to say Enterprise Conjur.

izgeri commented on docs-content/using.html.md.erb:
```diff
@@ -26,48 +30,90 @@ You can push the same application to multiple spaces. After preparing the applic
+  
+  For Conjur Enterprise (v4): 
+  
+  Add `!include my-app.yml` to a PCF policy file `pcf.yml` and reload the policy as follows:  
```
"to a PCF policy file `pcf.yml`" => "to an existing PCF policy file (in the example below, we use `pcf.yml`)"

perrygold replied:
done

izgeri commented on docs-content/using.html.md.erb:
```diff
@@ -26,48 +30,90 @@ You can push the same application to multiple spaces. After preparing the applic
+DB_USERNAME: !var pcf/my-app/db/username 
+DB_PASSWORD: !var pcf/my-app/db/password
+REGION: us-east-1
 SSL_CERT: !var:file ssl/certs/private
```
do you want the file to also be prefixed by `pcf` so it will be a secret stored in the `pcf` policy namespace?

perrygold replied:
yes. done

izgeri commented on docs-content/using.html.md.erb:
```diff
@@ -82,55 +128,62 @@ Add the instance-name from the previous step to the list of services in the appl
-The ID is stored in the `authn_login` string found under the VCAP_SERVICES environment variable. The `authn_login` string has the format `host/<id>`. For example:   
+The ID is stored in the `authn_login` string found under the VCAP_SERVICES environment variable. The `authn_login` string has the format `host/id`. For example:   
   
     `host/0299a19d-7de4-4e98-89f6-372ac7c0521f`
```
can we update this to `host/pcf/0299a19d-7de4-4e98-89f6-372ac7c0521f` for consistency with the rest of the docs?

perrygold replied:
done

perrygold replied:
BUT: I changed the prev. sentence to
The `authn_login` string has the format `host/namespace/binding_id`.

izgeri commented on docs-content/using.html.md.erb:
````diff
@@ -82,55 +128,62 @@ Add the instance-name from the previous step to the list of services in the appl
    ``` 
    - !grant
-     role: !group /path/to/some/app/secrets-users
+     role: !group /pcf/my-app/secrets-users
````
should just be "pcf/my-app/secrets-users" - it won't work with the first backslash there, I don't think

izgeri replied:
actually, if they are loading this directly into the `pcf` namespace, they don't need the `pcf` prefix at all.

perrygold replied:
changed to
   role: !group my-app/secrets-users

izgeri commented on docs-content/using.html.md.erb:
````diff
@@ -82,55 +128,62 @@ Add the instance-name from the previous step to the list of services in the appl
    ```
    
- If you save the policy to a file named `entitlements.yml`, you can load the policy into Conjur using the Conjur CLI. For example:  
+  If you save the above file as `entitlements.yml`, the following command loads it into the `pcf` namespace:  
````
If you are using Conjur version 5, you can save the above file as `entitlements.yml` ...
For version 4 Conjur, you can update your `pcf` policy to add this entitlement or to include an entitlements.yml file that contains:
```
- !grant
  role: !group my-app/secrets-users
  member: !host 0299a19d-7de4-4e98-89f6-372ac7c0521f
```

izgeri commented on docs-content/using.html.md.erb:
````diff
@@ -26,48 +30,90 @@ You can push the same application to multiple spaces. After preparing the applic
+  Add `!include my-app.yml` to a PCF policy file `pcf.yml` and reload the policy as follows:  
+  
+  ```
+  $ conjur policy load --as-group security_admin pcf.yml
````
If the `pcf.yml` file includes just the policies in the `pcf` namespace, this could actually be `--as-group pcf-admin-group`, because that's the group that owns the PCF policy. we do this in the demo.

perrygold commented on docs-content/index.html.md.erb:
```diff
@@ -69,11 +69,147 @@ The following table provides version and version-support information about Cyber
++ To configure the PCF tile, you must know the following information about your Conjur appliance:
+  * Conjur Account, which is specified when the Conjur appliance is created and configured. For hosted Conjur, the account is usually an email address that you specify on the web form.
+  * URL to the Conjur appliance.
+  * Valid Conjur user name for a user that has the ability to update and create PCF-related policy.   The Service Broker uses these credentials to add and remove host identities as you deploy and remove PCF applications. See [Recommendations](#recommendations).
```
I would support your adding a general Jira bug for this issue.  We are collecting backlogs such as this...
If you are suggesting that we add links over to our docs from the Pivotal tile doc, I want to  reject that idea. We've stated that a running Conjur is required, so they would have gotten to the docs by then.  Buried links are dangerous because they get old and point to the wrong thing, esp. if we are thinking of reorganizing everything.

izgeri replied:
happy to do that, though I'm not sure where to put it. please lmk if you know of a good place.

Also, can we change the language in this line to what I suggested above?

perrygold replied:
done

perrygold commented on docs-content/index.html.md.erb:
```diff
@@ -69,11 +69,147 @@ The following table provides version and version-support information about Cyber
+  * Valid Conjur user name for a user that has the ability to update and create PCF-related policy.   The Service Broker uses these credentials to add and remove host identities as you deploy and remove PCF applications. See [Recommendations](#recommendations).
+  *  Conjur API Key associated with the user name.
+  *  PCF Conjur Policy Namespace (optional). For production Conjur appliances, CyberArk strongly recommends  a dedicated PCF namespace;  otherwise, the Conjur root policy is the default.  See [Recommendations](#recommendations) for more information.
+  *  Conjur version. Either 4 for Enterprise Conjur or 5 for Open Source or hosted Conjur.
```
done

Commit:
Add script to run website locally in Docker

- Add Dockerfile and script to run site locally. Doesn't update on edit,
  unfortunately.
- Add note to README

Commit:
Update docs for 1.0 tile release

Commit:
Add upgrade instructions for 1.0.0

izgeri commented on docs-content/index.html.md.erb:
```diff
@@ -75,40 +74,47 @@ The following table provides version and version-support information about Cyber
-  * Conjur account, which is specified when the Conjur appliance is created and configured. For hosted Conjur, the account is usually an email address that you specify on the web form.
-  * URL to the Conjur appliance.
++ To configure the PCF tile, you must know the following information about your Conjur installation:
+  * Conjur account, which is specified when Conjur is installed and configured. For hosted Conjur, the account is usually an email address that you specify on the web form.
```
I still think it'd be nice to organize this information better - I wonder if a table would work?

micahlee replied:
I put the information in table. Let me know what you think!

izgeri commented on docs-content/installing.html.md.erb:
```diff
@@ -59,33 +63,38 @@ This tab configures communication between the Service Broker and a Conjur applia
   * **PCF Conjur Policy Branch ID**: The fully-qualified Conjur Policy ID of the dedicated Conjur policy for PCF that you created. Leave blank to default to the root policy (not recommended).
 
-  * **Conjur Version**: Defaults to 5. Select 4 for Enterprise Conjur. Select 5 for Open Source Conjur (including an evaluation Hosted Conjur).
+  * **Preserve Conjur Policy on Deprovision**: By default, when the service broker is removed
```
We should have a note here that this applies to Conjur EE v5+ and Conjur OSS v1+ only

izgeri commented on docs-content/using.html.md.erb:
```diff
@@ -110,26 +131,60 @@ In the command above:
 * For convenience, you may use the same instance-name in multiple PCF spaces, so that the same application manifest works in all spaces.
 
+**For Conjur Open Source and Conjur Enterprise v5:**
+Adding a service instance to a space will automatically create identities within Conjur
```
can we call these `resources` instead of `identities`?

izgeri commented on docs-content/using.html.md.erb:
```diff
@@ -110,26 +131,60 @@ In the command above:
+   # Entitle all applications in my-org
+   - !grant
+     role: !group my-app/secrets-users
+     member: !host /pcf/production/8f366b77-d826-4522-8cc6-261bf9b02cba
```
this should be `!layer` not `!host`, right?

izgeri commented on docs-content/using.html.md.erb:
```diff
@@ -110,26 +131,60 @@ In the command above:
+   # Entitle all applications in my-space
+   - !grant
+     role: !group my-app/secrets-users
+     member: !host /pcf/production/8f366b77-d826-4522-8cc6-261bf9b02cba/4b02b69c-ba35-4069-b634-ded5644740c9
```
same as above - `!layer`

izgeri commented on docs-content/index.html.md.erb:
```diff
@@ -3,67 +3,66 @@ title: CyberArk Conjur Service Broker for PCF (Beta)
     <tr>
         <td>Compatible Ops Manager version(s)</td>
-        <td>v1.11.x, v1.12.x, v2.0.x, v2.1.x, v2.2.x, and v2.3.x</td>
+        <td>v1.11.x, v1.12.x, v2.0.x, v2.1.x, v2.2.x, v2.3.x, v2.4.x, and v2.5.x</td>
```
Let's drop support for 1.11 here and below.

izgeri commented on docs-content/using.html.md.erb:
```diff
@@ -36,33 +36,54 @@ If your application uses another method to access secrets, such as Conjur API ca
+   pathname indicating the Conjur policy path. More information on the
+   format for `secrets.yml` [is available below](#secrets.yml).
 
 3. Store the `secrets.yml` file in the root folder of your application.
```
We should clarify here by adding something like

"If you deploy your application in a packaged format (eg as a JAR file), you will want to ensure your secrets.yml configuration file is available in the root application directory at deploy time (eg for a JAR file, you can store secrets.yml in the `src/main/resources` directory of your application.)"

micahlee replied:
Noting here before I forget. It would probably also be worthwhile to add something like this to the buildpack repo README.

izgeri commented on docs-content/using.html.md.erb:
```diff
@@ -110,26 +131,60 @@ In the command above:
+If you are going to have Conjur secrets only accesible to this application, use the
+`--no-start` option to avoid starting the application before it is configured to work
+with Conjur.
 
```
It's probably worth adding a notable note here that says something like

> If you're using Conjur EE v5+ or Conjur OSS v1+ and you're entitling your applications to access secrets in Conjur at the org or space level, then you're done!

Or maybe adding a heading that is like
> Optional additional steps for Conjur EE v4.x or entitling individual applications

izgeri commented:
Noting two things from XA:
- [x] our documentation should note that we ~recommend against~ do not allow [sharing service instances](https://docs.cloudfoundry.org/devguide/services/sharing-instances.html) in the interest of maintaining least privilege - users should provision a single service instance per space.
- [ ] ~we may need to document as a "known issue" that provisioning multiple service instances per space can cause problems with the org / space layer policies if you deprovision one and do not have "preserve policy" on [TBD]~
  - ~We will also need to write what error/indication to expect if you run into this issue, and how to resolve~
- [ ] our documentation should note that the 2.0.0 buildpack requires the secrets.yml to be in the app root dir at runtime, so Java apps using Spring Boot versions earlier than 1.4 will not be compatible 
- [x] We need to document the upgrade path from 0.3.3 to 1.0.0 (see [John's comment](https://github.com/cyberark/docs-cyberark-conjur-service-broker/pull/7#issuecomment-469715776) for more info)

micahlee commented on docs-content/installing.html.md.erb:
```diff
@@ -59,33 +63,38 @@ This tab configures communication between the Service Broker and a Conjur applia
   * **PCF Conjur Policy Branch ID**: The fully-qualified Conjur Policy ID of the dedicated Conjur policy for PCF that you created. Leave blank to default to the root policy (not recommended).
 
-  * **Conjur Version**: Defaults to 5. Select 4 for Enterprise Conjur. Select 5 for Open Source Conjur (including an evaluation Hosted Conjur).
+  * **Preserve Conjur Policy on Deprovision (Conjur EE v5+ and Conjur OSS v1+ only)**:
```
This option no longer exists.

jtuttle commented:
Upgrading from 0.3.3 to 1.0.0 works great if you rename both the service-broker and the service-broker app:

```
cf target -o cyberark-conjur-org
cf rename-service-broker conjur_service_broker conjur-service-broker
cf rename conjur_service_broker-0.3.3 conjur-service-broker-0.3.3
```

micahlee commented:
XA Notes from cyberark/conjur-service-broker#72

---

### Service broker index
- [x] Conjur API key - better description
- [x] Url conjur follower - better description
- [x] X509 description (nix paste)
- [x] V4 host factory note needs to be in the table as well

- Your-pfc-policy-file.yml
    - [x] Change comments from `/*` prefix to `#`
    - [x] Make comments about v4 full-line comments, not suffixed to the line
    - [x] Group ownership is wrong - needs to be absolute. This might be problematic if pcf policy isn't rooted in the root.
- Step 2
    - [x] bottom needs to note to save the host API key produced by loading the policy into conjur
- Step 3
    - [ ] ~Might need to be removed from overview and added to installation instructions~

### Installing and Configuring
- Service broker config
    - [ ] ~Recommendation: should be a new section before installing page~
    - [x] Appliance URL: Should say eval.conjur.org if it's a test instance
    - [x] Follower URL: Should mention that this is optional
    - [x] Conjur Login: should say that this id is from policy
    - [x] PCF Conjure Policy Branch ID: Use an example branch ID from loaded policy example
    - [x] Conjur Certificate: needs better info on how to get cert. Maybe use openssl s_client. Definitely sub conjur appliance configured with conjur cli configured
    - [x] Add click to save config

- [x] There is no stem cell tab - probably needs removing
- [x] Missing info after all the config you go to installation dashboard -> pending changes then click apply changes

### Using service broker
- [x] Step 1, Part 6 (new part) - Add info on how to set the variable value
- [x] Step 4: add info on creating a space to deploy the app to
- [x] Step 2: change "my-service-instance-name" to "conjur" (also match in step 3)
- [x] Step 2: formatting issue with CLI commands of cf and policy
- [x] Step 2: No showing on how to load the entitlement policy
- [x] Step 3: Include a link for more information on using the Conjur Buildpack

### Upgrade
- [x] I didn't locate any upgrade instructions in our docs. A good patter for this is the mysql upgrade docs: https://docs.pivotal.io/p-mysql/2-5/upgrade.html#upgrade-mysql-for-pcf

- [x] A future upgrade doc (and perhaps the release notes) MUST inform the user that with the new supply buildpack, the buildpack will need to be specified in either the manifest.yml or the cf push command. (And apps need to be restaged?)

### V4 Specific
 - [x] PivNet docs are missing policy load example for V4 entitlement.
 - [x] PivNet docs don't mention that you may need to install the policy plugin for the v4 CLI:
    ```sh-session
    $ conjur policy load /data/policy_v4/grant_host.yml
    error: Unknown command 'policy'
    ```
    ```sh-session
    $ conjur plugin install policy
    Successfully installed conjur-asset-policy-0.13.0
    1 gem installed
    ```

izgeri commented on docs-content/upgrading.html.md.erb:
```diff
@@ -0,0 +1,94 @@
+
+Version 1.0.0 also updates the Conjur Buildpack to version 2.0.0. This version changes the kind of
+buildpack from a decorator buildpack to a supply buildpack. This will likely require updating how
+your applications include and use the buildpack. More information about using the CyberArkConjur Buildpack
```
`CyberArkConjur` => `Conjur`?

micahlee replied:
To clarify, you want CyberArk removed, or is this just calling out the missing space?

izgeri replied:
I think generally we call it the `Conjur Buildpack`, so we should refer to it that way here too for consistency

izgeri commented on docs-content/upgrading.html.md.erb:
````diff
@@ -0,0 +1,94 @@
+    cf rename conjur_service_broker-0.3.3 conjur-service-broker-0.3.3
+    ```
+
+#### Updating Applications to use CyberArk Conjur Supply Buildpack
````
Does installing the new tile auto-delete the meta_buildpack and update the conjur_buildpack? Or is this also a manual step?

micahlee replied:
I know that uninstalling the tile does remove the buildpacks. I don't know what the behavior is for `0.3.3` to `1.0.0`. I suspect it would remove the meta_buildpack and replace the conjur_buildpack, but that needs to be tested to confirm one way or the other.

micahlee replied:
Note, the XA confirmed that the meta_buildpack does *NOT* get uninstalled during the upgrade. I've added a note to the XA feedback comment below that this should be included as an optional post-upgrade step.

micahlee commented:
@izgeri regarding:
> our documentation should note that the 2.0.0 buildpack requires the secrets.yml to be in the app root dir at runtime, so Java apps using Spring Boot versions earlier than 1.4 will not be compatible 

This feels like the kind of information that would get buried quickly in the current flows, what do you think about either:

- Adding a new page to the docs for Buildpack Usage that includes sections for language specific recommendations like this.

- Adding language/platform specific notes/recommendations to the buildpack GH README and linking to that for more language-specific guidance?

izgeri commented on docs-content/index.html.md.erb:
```diff
@@ -3,67 +3,66 @@ title: CyberArk Conjur Service Broker for PCF (Beta)
     <tr>
         <td>Release date</td>
-        <td>June 26, 2018</td>
+        <td>February 26, 2019</td>
```
March 6 maybe?

izgeri commented on docs-content/index.html.md.erb:
```diff
@@ -75,40 +74,122 @@ The following table provides version and version-support information about Cyber
+        <tr>
+            <td>
+                **x509 CA Certificate Chain for Conjur**<br/>
+                (Conjur EE v5.x+ and Conjur OSS v1.x+)<br/>
```
I think this is needed for any Conjur EE version; I expect Conjur OSS will also require this if Conjur is configured to use TLS, but I'm not sure what the workflow is to retrieve the certificate in that case

izgeri commented on docs-content/index.html.md.erb:
```diff
@@ -75,40 +74,122 @@ The following table provides version and version-support information about Cyber
+            <td>
+                The PEM encoded x509 certificate chain for the Conjur server. The certificate is
+                stored in a `.pem` file that was created by the `conjur init` command when using the
+                Conjur CLI. The default path for this file is `/conjur_install_directory/conjur-account.pem`.
```
I think elsewhere in our docs we're advising users retrieve the cert using `openssl` - see [here](https://docs.conjur.org/Latest/en/Content/Integrations/jenkins-configure.htm?Highlight=PEM) for example. Ideally there would be a conjur docs page we could link to, but I couldn't find anything just now that just has generic information about retrieving your certificate chain.

izgeri commented on docs-content/index.html.md.erb:
````diff
@@ -75,40 +74,122 @@ The following table provides version and version-support information about Cyber
 ```
 
-This can be loaded into the root policy or any other Conjur policy you would like. Host identities for PCF-deployed applications will be loaded into the `pcf` Conjur policy when the applications are bound to a Conjur service instance.
+This can be loaded into the root policy or any other Conjur policy you would like. Host identities for 
````
"or any other Conjur policy branch you would like" - I strongly prefer adding the `branch`, otherwise `policy` is an overloaded term

izgeri commented on docs-content/index.html.md.erb:
```diff
@@ -149,49 +230,79 @@ Create a policy using the following example as a guide. Save the policy as a YAM
+  needs to be permitted to read its own annotations.
 
-<pre class='term'>$ conjur host rotate_api_key -h pcf-service-broker</pre>
+* For **Conjur API Key**, use that host's API key. The admin who created the role can retrieve the value with this command:
```
Is it intentional that this remains a separate bullet? It looks like the bullet for `For Conjur Login` got merged into the bullet above

izgeri commented on docs-content/installing.html.md.erb:
Can we just say
> The tile will install the Conjur Buildpack, which will enable your applications to leverage Summon to automatically inject secrets. The default settings are appropriate.

yserota replied:
The tile **installs**...

izgeri commented on docs-content/release-notes.html.md.erb:
```diff
@@ -5,6 +5,26 @@ owner: Partners
+  functionality
+
+
+Known issues in this release: No known issues.
```
Should we add the buildpack limitation here?

> Conjur Buildpack does not support Java applications using Spring Boot 1.4+

izgeri commented on docs-content/using/deploy-app.html.md.erb:
```diff
@@ -0,0 +1,211 @@
+    > **NOTE:** If you deploy your application in a packaged format (e.g. as a JAR file),
+    > you will want to ensure your secrets.yml configuration file is available 
+    > in the root application directory at deploy time (e.g. for a JAR file, 
+    > you can store secrets.yml in the src/main/resources directory of your application).
```
Or here
> of your application if using Spring Boot versions below 1.4

(or similar)

izgeri commented on docs-content/installing.html.md.erb:
```diff
@@ -31,64 +31,93 @@ This adds the tile to your staging area.
 This tab configures communication between the Service Broker and a Conjur appliance.
 
-![Conjur Service Broker Configuration tab](service_broker_config_screen.png)
+![Conjur Service Broker Configuration tab](tile_config_screen.png)
```
this is going to need to be updated now that we don't have the "preserve policy" checkbox. it's just a screenshot of the tile config screen.

izgeri reviewed (changes requested):
I left a few comments, some of which will remain pending until we do the final XA tomorrow. Generally I think these are looking really great and I appreciate the effort you've put into them @micahlee! I know PM may have some feedback as well, I will share as I hear it.

izgeri commented on docs-content/upgrading.html.md.erb:
```diff
@@ -0,0 +1,95 @@
+---
+title: Updating CyberArk Conjur Service Broker for PCF
```
URL is `upgrading` but header says `updating`

yserota replied:
Upgrade or Update (no ing's :) )

izgeri commented on docs-content/index.html.md.erb:
````diff
@@ -201,13 +312,13 @@ If you decide to leave the **PCF Conjur Policy Namespace** blank in the tile con
+    ```
 
 ## <a id="limitations"></a>Limitations
 
````
the License section should have the copyright updated to 2019

micahlee commented:
Update upgrade docs to include:
- [x] How to re-provision existing spaces after upgrading to 1.0.0
    ```
    cf create-service cyberark-conjur community conjur-provision
    cf delete-service conjur-provision
    ```

- [x] How to re-bind existing apps to get new host ID with org/space permissions
   ```
    cf unbind-service [app-name] conjur
    cf bind-service [app-name] conjur
   ```

- [x] Remove meta buildback
   ```
   cf delete-buildback meta_buildpack
   ```

See: https://github.com/cyberark/conjur-service-broker/issues/72#issuecomment-470142198

yserota commented on docs-content/installing.html.md.erb:
...this is `https://eval.conjur.org`.

yserota commented on docs-content/installing.html.md.erb:
If the Host's API key changes, update this field and click **Apply Changes** to continue using the Service Broker to bind applications.

yserota commented on docs-content/release-notes.html.md.erb:
```diff
@@ -5,6 +5,26 @@ owner: Partners
+Features in this release:
+
+* Simplified policy management and deployment:
+  - Conjur Service Broker will automatically create policy branches in Conjur
```
Conjur Service Broker automatically creates policy branches in Conjur

yserota commented on docs-content/upgrading.html.md.erb:
```diff
@@ -0,0 +1,95 @@
+Upgrading the CyberArk Conjur Service Broker for PCF tile generally requires the same steps in
+the Ops Manager as when you first install the tile. Your configuration settings will migrate to
+the new version when you add the upgrade tile to the Ops Manager.
+
```
Upgrading the CyberArk Conjur Service Broker for the PCF tile requires the same steps  you ran in the Ops Manager when you first installed the tile.  Your configuration settings migrate to the new version when you add the upgrade tile to the Ops Manager.

yserota commented on docs-content/using/deploy-app.html.md.erb:
```diff
@@ -0,0 +1,211 @@
+
+This procedure enables a PCF application to obtain secrets and manage authorization privileges from an existing Conjur appliance.
+
+You can push the same application to multiple spaces. After preparing the application as described in Step 1 below, target your desired orgs and spaces and perform the remaining steps for each space.
```
After you prepare...

yserota commented on docs-content/using/deploy-app.html.md.erb:
```diff
@@ -0,0 +1,211 @@
+    > you can store secrets.yml in the src/main/resources directory of your application).
+
+4. If the secrets that your application needs are not already available in
+   Conjur policy, you can create a policy file `my-app.yml` that will add
```
Conjur plicy, create a policy file `my-app.yml` that adds

yserota commented on docs-content/using/deploy-app.html.md.erb:
```diff
@@ -0,0 +1,211 @@
+      the policy for resources and secrets outside of the service broker managed
+      policy branch.
+
+      Do define the resources to access `my-app.yml`, you would first define a policy
```
To define the resources to access `my-app.yml`,  define a policy...

yserota commented on docs-content/using/deploy-app.html.md.erb:
```diff
@@ -0,0 +1,211 @@
+
+##### Example: my-app.yml
+
+The following example shows a Conjur application policy declaring two secrets as variables for Conjur to manage and a group with access to those secrets.
```
replace "shows" with "describes"

yserota commented on docs-content/using/deploy-app.html.md.erb:
```diff
@@ -0,0 +1,211 @@
+
+#### <a id="permit-org-space"></a> Step 3: Permit Secret Access by Org and Space (Conjur Open Source and Conjur EE v5+ only)
+
+Adding a service instance to a space will automatically create resources within Conjur
```
Adding a service instance to a space  automatically creates resources within Conjur

yserota commented on docs-content/using/deploy-app.html.md.erb:
````diff
@@ -0,0 +1,211 @@
+4b02b69c-ba35-4069-b634-ded5644740c9
+```
+
+Assuming the Service Broker policy branch is `pcf/production`, the org and space layers
````
If  the Service Broker policy branch is `pcf/production`, the org and space layers

yserota commented on docs-content/using/deploy-app.html.md.erb:
````diff
@@ -0,0 +1,211 @@
+```
+
+Assuming the Service Broker policy branch is `pcf/production`, the org and space layers
+may be used to access variables in Conjur policy like:
````
may be used to access variables in Conjur policy.  For example:

yserota commented on docs-content/using/deploy-app.html.md.erb:
```diff
@@ -0,0 +1,211 @@
+If you're using Conjur EE v5+ or Conjur OSS v1+ and you're entitling your applications
+to access secrets in Conjur at the org or space level, then you're done!
+
+If you are going to have Conjur secrets entitled specifically to this application, use the
```
If you plan to have Conjur secrets...

yserota commented on docs-content/using/entitle-app.html.md.erb:
```diff
@@ -0,0 +1,82 @@
+
+> **NOTE:** These steps are optional for Conjur Open Source or Conjur EE v5+
+
+These steps allow you to entitle a specific application to access secrets in Conjur, rather
```
These steps enable you...

yserota commented on docs-content/using/entitle-app.html.md.erb:
```diff
@@ -0,0 +1,82 @@
+
+**For Conjur Open Source and Conjur Enterprise v5:** The application binding ID host is automatically
+added to the layers for the Organization and Space where the application is deployed. This means the
+app will inherit the permissions of the Org and Space without additional entitlements. These steps
```
app inherits the permissions

yserota reviewed (commented):
HI @izgeri  and @micahlee ,

Here are my comments:
- [ ]  Generally, I advise against using gerunds (ing words).  Make all headings active verbs. Install and configure
- [ ] Another general rule of thumb is to write in the present tense...update "will be" to "is" 

I made additional inline comments...none of which are show stoppers.  

Let me know if you have questions.

izgeri commented:
@yserota thanks!! I think many of the `ing` headings come straight from the Pivotal template, so some we may need to keep. Besides that, it looks like @micahlee has implemented most of your changes already :)

izgeri reviewed (approved):
LGTM!! 🍾 thanks for all your help with this @micahlee - it took a lot of effort, but the end result is a solid improvement!
